transl_utils/transl_utils.go

```diff
@@ -103,65 +147,259 @@ func TranslProcessGet(uriPath string, op *string) (*gnmipb.TypedValue, error) {
 	}
+	if rc.Auth.AuthEnabled {
+		req.AuthEnabled = true
+	}
+	resp, err1 := translib.Replace(req)
+
 	if err1 != nil{
-		log.V(2).Infof("REPLACE operation failed with error =%v", resp.ErrSrc)
-		return fmt.Errorf("REPLACE failed for this message")
+		log.V(2).Infof("REPLACE operation failed with error =%v, %v", resp.ErrSrc, err1.Error())
+		return err1
 	}
 
 
 	return nil
 }
 
 /* Update request handling. */
-func TranslProcessUpdate(uri string, t *gnmipb.TypedValue) error {
+func TranslProcessUpdate(uri string, t *gnmipb.TypedValue, ctx context.Context) error {
 	/* Form the CURL request and send to client . */
 	str := string(t.GetJsonIetfVal())
 	str3 := strings.Replace(str, "\n", "", -1)
 	log.V(2).Infof("Incoming JSON body is", str)
 
 	payload := []byte(str3)
-	req := translib.SetRequest{Path:uri, Payload:payload}
-	resp, err := translib.Create(req)
+	rc, ctx := common_utils.GetContext(ctx)
+	req := translib.SetRequest{Path:uri, Payload:payload, User: translib.UserRoles{Name: rc.Auth.User, Roles: rc.Auth.Roles}}
+	if rc.BundleVersion != nil {
+		nver, err := translib.NewVersion(*rc.BundleVersion)
+		if err != nil {
+			log.V(2).Infof("UPDATE operation failed with error =%v", err.Error())
+			return err
+		}
+		req.ClientVersion = nver
+	}
+	if rc.Auth.AuthEnabled {
+		req.AuthEnabled = true
+	}
+	resp, err := translib.Update(req)
 	if err != nil{
-		//If Create fails, it may be due to object already existing/can not be created
-		// such as interface, in this case use Update.
-		resp, err = translib.Update(req)
+		switch err.(type) {
+		case tlerr.NotFoundError:
+			//If Update fails, it may be due to object not existing in this case use Replace to create and update the object.
+			resp, err = translib.Replace(req)
+		default:
+			log.V(2).Infof("UPDATE operation failed with error =%v, %v", resp.ErrSrc, err.Error())
+			return err
+		}
 	}
 	if err != nil{
-		log.V(2).Infof("UPDATE operation failed with error =%v", resp.ErrSrc)
-		return fmt.Errorf("UPDATE failed for this message")
+		log.V(2).Infof("UPDATE operation failed with error =%v, %v", resp.ErrSrc, err.Error())
+		return err
 	}
 	return nil
 }
 
+func TranslProcessBulk(delete []*gnmipb.Path, replace []*gnmipb.Update, update []*gnmipb.Update, prefix *gnmipb.Path, ctx context.Context) error {
+	var br translib.BulkRequest
+	var uri string
+
+        var deleteUri []string
+        var replaceUri []string
+        var updateUri []string
+
+	rc, ctx := common_utils.GetContext(ctx)
+	log.V(2).Info("TranslProcessBulk Called")
+	for _,d := range delete {
+		ConvertToURI(prefix, d, &uri)
+		var str3 string
+		payload := []byte(str3)
+		req := translib.SetRequest{
+			Path: uri,
+			Payload: payload,
+			User: translib.UserRoles{Name: rc.Auth.User, Roles: rc.Auth.Roles},
+		}
+		if rc.BundleVersion != nil {
+			nver, err := translib.NewVersion(*rc.BundleVersion)
+			if err != nil {
+				log.V(2).Infof("Bulk Set operation failed with error =%v", err.Error())
+				return err
```

Comment:
I changed the message to "Bundle Version Check failed with error =%v" but you are right, currently all set failures return "Unknown" error code since we are just directly returning the translib error object. Instead we should be translating these errors into appropriate gRPC error codes. Not sure if we should address this in a different PR though.